components/org.wso2.carbon.identity.oauth.ciba/src/main/java/org/wso2/carbon/identity/oauth/ciba/dao/CibaAuthMgtDAOImpl.java

```diff
@@ -0,0 +1,547 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth.ciba.dao;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.oauth.ciba.exceptions.CibaCoreException;
+import org.wso2.carbon.identity.oauth.ciba.exceptions.ErrorCodes;
+import org.wso2.carbon.identity.oauth.ciba.model.CibaAuthCodeDO;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Implementation of abstract DAO layer.
+ */
+public class CibaAuthMgtDAOImpl implements CibaAuthMgtDAO {
+
+    private static final Log log = LogFactory.getLog(CibaAuthMgtDAOImpl.class);
+
+    private CibaAuthMgtDAOImpl() {
+
+    }
+
+    private static CibaAuthMgtDAOImpl cibaAuthMgtDAOImplImplInstance = new CibaAuthMgtDAOImpl();
+
+    static CibaAuthMgtDAOImpl getInstance() {
+
+        if (cibaAuthMgtDAOImplImplInstance == null) {
+
+            synchronized (CibaAuthMgtDAOImpl.class) {
+
+                if (cibaAuthMgtDAOImplImplInstance == null) {
+
+                    /* instance will be created at request time */
+                    cibaAuthMgtDAOImplImplInstance = new CibaAuthMgtDAOImpl();
+                }
+            }
+        }
+        return cibaAuthMgtDAOImplImplInstance;
+
+    }
+
+    /**
+     * Persists the status of the relevant CibAuthCode identified by the CibaAuthCodeDOKey.
+     *
+     * @param cibaAuthCodeDOKey       Identifier for CibaAuthCodeDOKey.
+     * @param cibaAuthentcationStatus Status of the relevant Ciba Authentication.
+     * @throws CibaCoreException Exception thrown from CibaCore Component.
+     */
+    public void persistStatus(String cibaAuthCodeDOKey, String cibaAuthentcationStatus) throws CibaCoreException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(true)) {
+            try (PreparedStatement prepStmt = connection.prepareStatement(SQLQueries.
+                    CibaSQLQueries.UPDATE_AUTHENTICATION_STATUS)) {
+
+                prepStmt.setString(1, cibaAuthentcationStatus);
+                prepStmt.setString(2, cibaAuthCodeDOKey);
+
+                prepStmt.execute();
+                connection.commit();
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Successfully persisted the authentication_status identified by AuthCodeDOKey : " +
+                            cibaAuthCodeDOKey);
+                }
+            }
+        } catch (SQLException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error occurred in persisting the authentication_status identified by AuthCodeDOKey : " +
+                        cibaAuthCodeDOKey);
+            }
+            throw new CibaCoreException(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
```

Comment:
Why do we need to throw HTTP status codes from the OSGi layer?

Reply:
It is an internal error. It is to make error response so specific. If not when we capture all Ciba core exceptions might be difficult what kind of error [internal, bad request error] occurred.

Reply:
Then it would be better to throw two types of exceptions like Client and Server exception with a set of defined error codes to distinguish among the same type.

Reply:
Refactored. The status will be set in the endpoint. CibaCore Exception will be always an internal error. So endpoint will set the status for that. Other types of errors will be AuthFailure Exception to mark error codes related to authentication.